src/Compilers/CSharp/Test/Semantic/Semantics/BindingAsyncTasklikeMoreTests.cs

```diff
@@ -927,5 +937,316 @@ static void Main()
+        /// </summary>
+        [WorkItem(21500, "https://github.com/dotnet/roslyn/issues/21500")]
+        [Fact]
+        public void AwaitUnsafeOnCompleted_AdditionalConstraintMissingFromAwaiter()
+        {
+            var source0 =
+@"using System;
+using System.Runtime.CompilerServices;
+namespace System.Runtime.CompilerServices
+{
+    class AsyncMethodBuilderAttribute : Attribute
+    {
+        public AsyncMethodBuilderAttribute(Type t) { }
+    }
+}
+public interface IMyAwaiter { }
+[AsyncMethodBuilder(typeof(MyTaskMethodBuilder))]
+public sealed class MyTask
+{
+    public Awaiter GetAwaiter() => new Awaiter();
+    public class Awaiter : ICriticalNotifyCompletion
+    {
+        public void OnCompleted(Action a) { }
+        public void UnsafeOnCompleted(Action a) { }
+        public bool IsCompleted => true;
+        public void GetResult() { }
+    }
+}
+public sealed class MyTaskMethodBuilder
+{
+    public static MyTaskMethodBuilder Create() => new MyTaskMethodBuilder();
+    public void SetStateMachine(IAsyncStateMachine stateMachine) { }
+    public void Start<TStateMachine>(ref TStateMachine stateMachine)
+        where TStateMachine : IAsyncStateMachine
+    {
+    }
+    public void SetException(Exception e) { }
+    public void SetResult() { }
+    public void AwaitOnCompleted<TAwaiter, TStateMachine>(ref TAwaiter awaiter, ref TStateMachine stateMachine)
+        where TAwaiter : INotifyCompletion
+        where TStateMachine : IAsyncStateMachine
+    {
+    }
+    public void AwaitUnsafeOnCompleted<TAwaiter, TStateMachine>(ref TAwaiter awaiter, ref TStateMachine stateMachine)
+        where TAwaiter : IMyAwaiter, ICriticalNotifyCompletion
+        where TStateMachine : IAsyncStateMachine
+    {
+    }
+    public MyTask Task => new MyTask();
+}";
+            var compilation0 = CreateCompilationWithMscorlib45(source0);
+            var ref0 = compilation0.EmitToImageReference();
+            var source =
+@"class Program
+{
+    static async MyTask F()
+    {
+        await new MyTask();
+    }
+    static void Main()
+    {
+        var t = F();
+        t.GetAwaiter().GetResult();
+    }
+}";
+            var compilation = CreateCompilationWithMscorlib45(source, references: new[] { ref0 });
+            compilation.VerifyEmitDiagnostics(
+                // (5,9): error CS0311: The type 'MyTask.Awaiter' cannot be used as type parameter 'TAwaiter' in the generic type or method 'MyTaskMethodBuilder.AwaitUnsafeOnCompleted<TAwaiter, TStateMachine>(ref TAwaiter, ref TStateMachine)'. There is no implicit reference conversion from 'MyTask.Awaiter' to 'IMyAwaiter'.
+                //         await new MyTask();
+                Diagnostic(ErrorCode.ERR_GenericConstraintNotSatisfiedRefType, "await new MyTask();").WithArguments("MyTaskMethodBuilder.AwaitUnsafeOnCompleted<TAwaiter, TStateMachine>(ref TAwaiter, ref TStateMachine)", "IMyAwaiter", "TAwaiter", "MyTask.Awaiter").WithLocation(5, 9));
+        }
```

Comment:
Is it useful to also test case where Awaiter doesn't inherit INotifyCompletion?

Reply:
Added.